### DISCUSSION TITLE
Predefined WFS ETS fails on AutoConfigure WFS-GetNumber of Feautures in DB

### DISCUSSION THREAD

**Discussion by klimeto** [INSPIRE-MIF/helpdesk-validator#904] [Q&A] [answered]
Dears,

While testing this WFS `https://arc.sazp.sk/arcgis/services/tazobny_odpad/ulozisko__pd3801/MapServer/WFSServer?SERVICE=WFS&REQUEST=GetCapabilities` in production ETF Web App `http://inspire.ec.europa.eu/validator` I failed with the error on `Get the number of available Feature Types with a GetFeature request with RESULTTYPE=hits. Called by 'Analyze WFS Capabilities'.`

By checking the request sent by ETF `https://arc.sazp.sk/arcgis/services/tazobny_odpad/ulozisko__pd3801/MapServer/WFSServer?OUTPUTFORMAT=text/xml;%20subtype=gml/3.2&REQUEST=GetFeature&RESULTTYPE=hits&SERVICE=WFS&TYPENAMES=PD3801_Extractive_waste:ULOZISKO_VSETKY&VERSION=2.0.0` the service responds with extension, which is indeed correct since for `GetFeatures` request the following output formats are advertised in GetCapabilities:

```xml
<ows:AllowedValues>
<ows:Value>GML32</ows:Value>
<ows:Value>GML32+ZIP</ows:Value>
<ows:Value>application/gml+xml; version=3.2</ows:Value>
<ows:Value>GML3</ows:Value>
<ows:Value>GML3+ZIP</ows:Value>
<ows:Value>text/xml; subtype=gml/3.1.1</ows:Value>
<ows:Value>GML2</ows:Value>
<ows:Value>GML2+ZIP</ows:Value>
<ows:Value>text/xml; subtype=gml/2.1.2</ows:Value>
<ows:Value>GEOJSON</ows:Value>
<ows:Value>GEOJSON+ZIP</ows:Value>
<ows:Value>ESRIGEOJSON</ows:Value>
<ows:Value>ESRIGEOJSON+ZIP</ows:Value>
<ows:Value>KML</ows:Value>
<ows:Value>application/vnd.google-earth.kml xml</ows:Value>
<ows:Value>application/vnd.google-earth.kml+xml</ows:Value>
<ows:Value>KMZ</ows:Value>
<ows:Value>application/vnd.google-earth.kmz</ows:Value>
<ows:Value>SHAPE+ZIP</ows:Value>
<ows:Value>CSV</ows:Value>
<ows:Value>CSV+ZIP</ows:Value>
</ows:AllowedValues>
```

and the one ETF used is not advertised.

If you instead use `GML32` as you actually already do in the previous test assertion 'GetFeatures' the service would respond correctly:
```url
https://arc.sazp.sk/arcgis/services/tazobny_odpad/ulozisko__pd3801/MapServer/WFSServer?OUTPUTFORMAT=GML32&REQUEST=GetFeature&RESULTTYPE=hits&SERVICE=WFS&TYPENAMES=PD3801_Extractive_waste:ULOZISKO_VSETKY&VERSION=2.0.0
```
```xml
<wfs:FeatureCollection xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:wfs="http://www.opengis.net/wfs/2.0" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:PD3801_Extractive_waste="https://arc.sazp.sk/arcgis/services/tazobny_odpad/ulozisko__pd3801/MapServer/WFSServer" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" timeStamp="2020-02-27T21:07:31Z" numberMatched="455" numberReturned="0" xsi:schemaLocation="http://www.opengis.net/wfs/2.0 http://schemas.opengis.net/wfs/2.0/wfs.xsd http://www.opengis.net/gml/3.2 http://schemas.opengis.net/gml/3.2.1/gml.xsd https://arc.sazp.sk/arcgis/services/tazobny_odpad/ulozisko__pd3801/MapServer/WFSServer https://arc.sazp.sk/arcgis/services/tazobny_odpad/ulozisko__pd3801/MapServer/WFSServer?service=wfs%26version=2.0.0%26request=DescribeFeatureType"> </wfs:FeatureCollection>
```

Could you please clarify why this ETS assertion uses OUTPUTFORMAT parameter value which is not advertised in the Capabilities of the service?

The URL to ETF report is here:
```url
http://inspire.ec.europa.eu/validator/v2/TestRuns/EIDc230f334-6be6-4d09-b49f-afbbc249d8ef.html#EID8867f704-0170-1000-3737-56b08aaa553c
```
Thank you,
Tomas

`

**Comment by iuriemaxim** [comment]
Hello Tomas,

Indeed it seems that the request made by the ETF validator is not correct. It should require the features with the following MIME type: "**aplication/gml+xml; version=3.2**" according to Technical Guidelines that requiers features to be provided in GML 3.2 and according to IANA (https://www.iana.org/assignments/media-types/application/gml+xml). And "aplication/gml+xml; version=3.2"  is present in the GetCapabilities response.

In any case "**text/xml; subtype=gml/3.2**" should not be correct for GetFeatures request. Neither "**application/gml+xml; subtype=3.2**" should not be correct if consulting IANA at https://www.iana.org/assignments/media-types/application/gml+xml.

But I think you should look in more details to the GetCapabilities response as there are some more issues and I think that the error you get is from another reason. The error you mention indicates that the validator is not able to get the number of available feature types (not the number of features of each feature type). And most probably this is due to the GetCapabilities document that is not complete or incorrect.

According to the WFS standard 2.0.0/2.0.2 the MIME type for GetCapabilities request shall include the value ”text/xml” and may include other MIME types as well. 

The WFS 2.0.2 standard can be consulted here: http://docs.opengeospatial.org/is/09-025r2/09-025r2.html

Below an extract on the topic:

![image](https://user-images.githubusercontent.com/20813598/75668895-b1feef80-5c82-11ea-9ffa-6e80585a27c2.png)

Unfortunately the service tested does not include ”text/xml”. It should look like this:

![image](https://user-images.githubusercontent.com/20813598/75669205-508b5080-5c83-11ea-8b30-40f33a5e5604.png)

And your service looks like this and is missing AcceptFormats and Sections parameters for GetCapabilities request:

![image](https://user-images.githubusercontent.com/20813598/75669459-c0014000-5c83-11ea-9dac-7c1bc1ece885.png).

You may find an example of a WFS that is passing the ETF tests:

https://inspire.meteoromania.ro/WIGOS/WFS?service=WFS&request=GetCapabilities

The issues that are reported by the validator for the above service are fake and they were already communicated and are treated by the development team.

You may look at it and compare to your implementation.


Regarding FeatureTypes, your GetCapabilities document looks like this:

![image](https://user-images.githubusercontent.com/20813598/75672582-b549a980-5c89-11ea-888d-f664fbcc0e99.png)

while an example that is passing the test looks like this:

![image](https://user-images.githubusercontent.com/20813598/75672669-e3c78480-5c89-11ea-96e1-24c67bbf9ac5.png)

"PD3801_Extractive_waste:" is not a correct prefix for an INSPIRE data theme.

INSPIRE prefixes and schemas are found here:

https://inspire.ec.europa.eu/schemas/

Most probably the error message provided by the validator should be different and the OUTPUTFORMAT parameter of the GetFeatures request should be changed into "**aplication/gml+xml; version=3.2**".

Hope it helps,
Iurie

**Comment by iuriemaxim** [comment]
Most probably your data is falling under Production and Industrial Facilities Data Theme.

Technical Guideline is here:
https://inspire.ec.europa.eu/id/document/tg/pf

INSPIRE schema is here:
https://inspire.ec.europa.eu/schemas/pf/4.0/ProductionAndIndustrialFacilities.xsd

Prefix would be then "pf".

Seems like your GetCapabilities document is listing Features instead of listing FeatureTypes.

You will have only one feature type and that would be "**pf:ProductionAndIndustrialFacilities**"

**Comment by iuriemaxim** [comment]
@carlospzurita are there any updates related to the issue that the validator is not requesting the features with "aplication/gml+xml; version=3.2" MIME type, according to Technical Guidelines that requires features to be provided in GML 3.2 and according to IANA (https://www.iana.org/assignments/media-types/application/gml+xml)?

**Comment by iuriemaxim** [comment, marked as answer]
@klimeto The media types encircled in red are not listed in IANA

![image](https://user-images.githubusercontent.com/20813598/95634584-ea757000-0a92-11eb-9a72-792191934db2.png)

For geojson media type, the correct syntax is found here https://www.iana.org/assignments/media-types/application/geo+json

Entire list of accepted media types are found here: https://www.iana.org/assignments/media-types/media-types.xhtml

None of those encircled in red are listed in IANA.

While testing against predefined WFS it shows this error

![image](https://user-images.githubusercontent.com/20813598/95636336-8bfec080-0a97-11eb-8741-ba4fc214d8bd.png)

The error is due to this declaration, while it should be application/gml+xml; version=3.2

![image](https://user-images.githubusercontent.com/20813598/95636490-f6affc00-0a97-11eb-911c-d9cd5091cd58.png)

as it is declared here 
![image](https://user-images.githubusercontent.com/20813598/95636669-632afb00-0a98-11eb-91b5-09a8b77f2ee8.png)

Once this error will be fixed the WFS will be able to provide the features and the mentioned error will disappear as well:
![image](https://user-images.githubusercontent.com/20813598/95636956-2f9ca080-0a99-11eb-936b-2c81aedf92fd.png)

Just paste the link indicated in the error message in any browser (https://arc.sazp.sk/arcgis/services/tazobny_odpad/ulozisko__pd3801/MapServer/WFSServer?#PROJECT#ENDPT.GETFEATURE.GET=%23Project%23endpt.GetFeature.Get&NAMESPACES=xmlns%28PD3801_Extractive_waste%2Chttps%3A%2F%2Farc.sazp.sk%2Farcgis%2Fservices%2Ftazobny_odpad%2Fulozisko__pd3801%2FMapServer%2FWFSServer%29&OUTPUTFORMAT=text%2Fxml%3B+subtype%3Dgml%2F3.2&REQUEST=GetFeature&RESULTTYPE=hits&SERVICE=WFS&TYPENAMES=PD3801_Extractive_waste%3AULOZISKO_VSETKY&VERSION=2.0.0)

and will show

![image](https://user-images.githubusercontent.com/20813598/95637121-a043bd00-0a99-11eb-9a40-72efc94243fe.png)

The request is made based on this declaration
![image](https://user-images.githubusercontent.com/20813598/95636490-f6affc00-0a97-11eb-911c-d9cd5091cd58.png)

not based on this declaration 
![image](https://user-images.githubusercontent.com/20813598/95636669-632afb00-0a98-11eb-91b5-09a8b77f2ee8.png)


This is quite similar to https://github.com/inspire-eu-validation/community/issues/29

@carlospzurita This can be marked as discussion. The validator provides all the necessary information for the user to fix the issue.

**Comment by fabiovinci** [comment]
Dear @klimeto,

do you still have this kind of issue?

Have you tried to follow the instructions provided by iuriemaxim to solve the problem?

**Comment by iuriemaxim** [comment]
@klimeto 

Did ESRI fixed this issue?

Probably relevant for the subject:
https://inspire.ec.europa.eu/media-types/application/application.en.xml
